src/plugins/nvdimm.c

```diff
@@ -0,0 +1,549 @@
+    gint ret = 0;
+
+    ret = ndctl_new (&ctx);
+    if (ret != 0) {
+        g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                     "Failed to create ndctl context");
+        return FALSE;
+    }
+
+    ndns = get_namespace_by_name (namespace, ctx);
+    if (!ndns) {
+        g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_NOEXIST,
+                     "Failed to disable namespace: namespace '%s' not found.", namespace);
+        return FALSE;
+    }
+
+    ret = ndctl_namespace_disable_safe (ndns);
+    if (ret != 0) {
+        g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                     "Failed to disable namespace: %s", strerror (-ret));
+        ndctl_unref (ctx);
+        return FALSE;
+    }
+
+    ndctl_unref (ctx);
+    return TRUE;
+}
+
+static BDNVDIMMNamespaceInfo* get_nvdimm_namespace_info (struct ndctl_namespace *ndns, GError **error) {
+    struct ndctl_btt *btt;
+    struct ndctl_pfn *pfn;
+    struct ndctl_dax *dax;
+    enum ndctl_namespace_mode mode;
+    gchar buf[37] = {0};
+    uuid_t uuid;
+
+    btt = ndctl_namespace_get_btt (ndns);
+    dax = ndctl_namespace_get_dax (ndns);
+    pfn = ndctl_namespace_get_pfn (ndns);
+    mode = ndctl_namespace_get_mode (ndns);
+
+    BDNVDIMMNamespaceInfo *info = g_new0 (BDNVDIMMNamespaceInfo, 1);
+
+    info->dev = g_strdup (ndctl_namespace_get_devname (ndns));
+
+    switch (mode) {
+        case NDCTL_NS_MODE_MEMORY:
+            if (pfn)
+              info->size = ndctl_pfn_get_size (pfn);
+            else
+              info->size = ndctl_namespace_get_size (ndns);
+          info->mode = BD_NVDIMM_NAMESPACE_MODE_MEMORY;
+          break;
+        case NDCTL_NS_MODE_DAX:
+            if (!dax) {
+                g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                             "Failed to get information about namespaces.");
+                bd_nvdimm_namespace_info_free (info);
+                return NULL;
+            }
+            info->size = ndctl_dax_get_size (dax);
+            info->mode = BD_NVDIMM_NAMESPACE_MODE_DAX;
+            break;
+        case NDCTL_NS_MODE_SAFE:
+            if (!btt) {
+                g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                             "Failed to get information about namespaces.");
+                bd_nvdimm_namespace_info_free (info);
+                return NULL;
+            }
+            info->size = ndctl_btt_get_size (btt);
+            info->mode = BD_NVDIMM_NAMESPACE_MODE_SECTOR;
+            break;
+        case NDCTL_NS_MODE_RAW:
+            info->size = ndctl_namespace_get_size (ndns);
+            info->mode = BD_NVDIMM_NAMESPACE_MODE_RAW;
+            break;
+        default:
+            g_set_error (error, BD_NVDIMM_ERROR, BD_NVDIMM_ERROR_NAMESPACE_FAIL,
+                         "Failed to get information about namespaces.");
```

Comment:
I think that adding the mode to the error message is going to simplify debugging one day.